// cardano-js/dist/Address/JormungandrAddress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const JORMUNGANDR_PUBKEY_LENGTH = 32;
var JormungandrAddressKind;
(function (JormungandrAddressKind) {
    JormungandrAddressKind[JormungandrAddressKind["singleAddress"] = 3] = "singleAddress";
    JormungandrAddressKind[JormungandrAddressKind["groupedAddress"] = 4] = "groupedAddress";
    JormungandrAddressKind[JormungandrAddressKind["accountAddress"] = 5] = "accountAddress";
    JormungandrAddressKind[JormungandrAddressKind["multisigAddress"] = 6] = "multisigAddress";
})(JormungandrAddressKind = exports.JormungandrAddressKind || (exports.JormungandrAddressKind = {}));
var JormungandrEnvironment;
(function (JormungandrEnvironment) {
    JormungandrEnvironment["production"] = "production";
    JormungandrEnvironment["testing"] = "testing";
})(JormungandrEnvironment = exports.JormungandrEnvironment || (exports.JormungandrEnvironment = {}));
var JormungandrAddressName;
(function (JormungandrAddressName) {
    JormungandrAddressName["single"] = "single";
    JormungandrAddressName["grouped"] = "grouped";
    JormungandrAddressName["account"] = "account";
    JormungandrAddressName["multisig"] = "multisig";
})(JormungandrAddressName = exports.JormungandrAddressName || (exports.JormungandrAddressName = {}));
function translateJormungandrAddressKind(kind) {
    switch (kind) {
        case 3:
            return JormungandrAddressName.single;
        case 4:
            return JormungandrAddressName.grouped;
        case 5:
            return JormungandrAddressName.account;
        case 6:
            return JormungandrAddressName.multisig;
        default:
            throw new Error('Unrecognised JormungandrAddressKind');
    }
}
exports.translateJormungandrAddressKind = translateJormungandrAddressKind;
function decodeJormungandrAddress(bytes) {
    const kind = bytes[0] & 0b01111111;
    const environment = bytes[0] & 0b10000000
        ? JormungandrEnvironment.testing
        : JormungandrEnvironment.production;
    switch (kind) {
        case JormungandrAddressKind.singleAddress:
            if (bytes.byteLength !== 1 + JORMUNGANDR_PUBKEY_LENGTH) {
                return null;
            }
            break;
        case JormungandrAddressKind.groupedAddress:
            if (bytes.byteLength !== 1 + 2 * JORMUNGANDR_PUBKEY_LENGTH) {
                return null;
            }
            break;
        case JormungandrAddressKind.accountAddress:
            if (bytes.byteLength !== 1 + JORMUNGANDR_PUBKEY_LENGTH) {
                return null;
            }
            break;
        case JormungandrAddressKind.multisigAddress:
            if (bytes.byteLength !== 1 + JORMUNGANDR_PUBKEY_LENGTH) {
                return null;
            }
            break;
        default:
            return null;
    }
    return { addressName: translateJormungandrAddressKind(kind), environment };
}
exports.decodeJormungandrAddress = decodeJormungandrAddress;
