// cardano-js/dist/Address/__test__/data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addresses = {
    Byron: {
        Mainnet: {
            Daedalus: 'DdzFFzCqrhsw3prhfMFDNFowbzUku3QmrMwarfjUbWXRisodn97R436SHc1rimp4MhPNmbdYb1aTdqtGSJixMVMi5MkArDQJ6Sc1n3Ez',
            Icarus: 'Ae2tdPwUPEZFRbyhz3cpfC2CumGzNkFBN2L42rcUc2yjQpEkxDbkPodpMAi',
        },
        Testnet: {
            Daedalus: '37btjrVyb4KEB2STADSsj3MYSAdj52X5FrFWpw2r7Wmj2GDzXjFRsHWuZqrw7zSkwopv8Ci3VWeg6bisU9dgJxW5hb2MZYeduNKbQJrqz3zVBsu9nT',
            Icarus: '2cWKMJemoBakkUSWX3DZdx8eXGeqjN6mkgVUCND1RNB736qWS5v1CGgQGNNTFUZSXaVLj',
        },
        invalid: {
            short: 'EkxDbkPo',
            networkMagic: '3reY92cShRkjtmz7q31547czPNHbrhbRGhVLehTrNDNDNeDaKJwcM8aMmWg2zd7cHVFvhdui4a86nEdsSEE7g7kcZKKvBw7nzixnbX1',
        },
    },
    Jormungandr: {
        Mainnet: {
            single: 'ca1qdaa2wrvxxkrrwnsw6zk2qx0ymu96354hq83s0r6203l9pqe6677zqx4le2',
            grouped: 'ca1q3aa2wrvxxkrrwnsw6zk2qx0ymu96354hq83s0r6203l9pqe6677rvjwwzcv9nhtynxf728nserccua2w8q949dqzxdmj8wcazwrty4wga8haz',
        },
        Testnet: {
            single: 'ta1s00e7zr89gafgauz9xu3m25cz5ugs0s4xhtxdhqsuca58r6ycclr7v3je63',
            grouped: 'ta1s3aa2wrvxxkrrwnsw6zk2qx0ymu96354hq83s0r6203l9pqe6677rvjwwzcv9nhtynxf728nserccua2w8q949dqzxdmj8wcazwrty4we4spcz',
        },
        invalid: {
            single: 'ca1qvqsyqcyq5rqwzqfpg9scrgk66qs0',
            wrongKind: 'ca1dvqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jqscdket',
        },
    },
};
